src/Compilers/CSharp/Test/Emit/Emit/LocalStateTracing/LocalStateTracingTests.cs

```diff
@@ -0,0 +1,5898 @@
+    }
+}
+""";
+
+        private static string WithHelpers(string source)
+            => source + s_helpers;
+
+        private static readonly TargetFramework s_targetFramework = TargetFramework.Net70;
+
+        private static readonly Verification s_verification = Verification.Fails with
+        {
+            ILVerifyMessage = """
+            [LogMethodEntry]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x9 }
+            [LogLambdaEntry]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x9 }
+            [LogStateMachineMethodEntry]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x8 }
+            [LogStateMachineLambdaEntry]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x8 }
+            [Entry]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0xbc }
+            """
+        };
+
+        private CompilationVerifier CompileAndVerify(string source, string? ilVerifyMessage = null, string? expectedOutput = null)
+            => CompileAndVerify(
+                source,
+                options: (expectedOutput != null) ? TestOptions.UnsafeDebugExe : TestOptions.UnsafeDebugDll,
+                emitOptions: s_emitOptions,
+                verify: s_verification with { ILVerifyMessage = ilVerifyMessage + Environment.NewLine + s_verification.ILVerifyMessage },
+                targetFramework: s_targetFramework,
+                expectedOutput: expectedOutput);
+
+        private static void AssertNotInstrumented(CompilationVerifier verifier, string qualifiedMethodName)
+            => AssertInstrumented(verifier, qualifiedMethodName, expected: false);
+
+        private static void AssertInstrumented(CompilationVerifier verifier, string qualifiedMethodName, bool expected = true)
+        {
+            var il = verifier.VisualizeIL(qualifiedMethodName);
+            var isInstrumented = il.Contains(TrackerTypeName);
+
+            AssertEx.AreEqual(expected, isInstrumented,
+                $"Method '{qualifiedMethodName}' should {(expected ? "be" : "not be")} instrumented. Actual IL:{Environment.NewLine}{il}");
+        }
+
+        [Fact]
+        public void Composition_LocalStateTracing_TestCoverage()
+        {
+            var source = WithHelpers(@"
+class C
+{
+    static void Main(string[] p)
+    {
+        string[] a = p = null;
+        Microsoft.CodeAnalysis.Runtime.Instrumentation.FlushPayload();
+    }
+}
+") + CSharpInstrumentationChecker.InstrumentationHelperSource;
+
+            var verifier = CompileAndVerify(
+                source,
+                options: TestOptions.UnsafeDebugExe,
+                emitOptions: EmitOptions.Default
+                    .WithInstrumentationKinds(ImmutableArray.Create(InstrumentationKind.LocalStateTracing, InstrumentationKind.TestCoverage)),
+                verify: s_verification,
+                targetFramework: s_targetFramework);
```

Comment:
Can we verify expected output here? The local state tracker should write to console, right? Same in `Composition_LocalStateTracing_TestCoverage_LocallySuppressed` below.

Reply:
The LST instrumenter will instrument the test helpers of code coverage instrumenter producing quite a bit of output. Not sure if there is value in capturing that.